### PR TITLE
Benchmarks: Add Benchmark - Add ib traffic validation distributed benchmark

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
Add ib traffic validation distributed benchmark.

**Major Revision**
- Add ib traffic validation distributed benchmark, example and test

### CODE REVIEW

File: superbench/benchmarks/micro_benchmarks/__init__.py [hunk 1]
```diff
@@ -19,10 +19,11 @@
...
 from superbench.benchmarks.micro_benchmarks.rocm_gemm_flops_performance import RocmGemmFlopsBenchmark
 from superbench.benchmarks.micro_benchmarks.gpu_sm_copy_bw_performance import GpuSmCopyBwBenchmark
+from superbench.benchmarks.micro_benchmarks.ib_validation_performance import IBTrafficBenchmark
 
 __all__ = [
     'MicroBenchmark', 'MicroBenchmarkWithInvoke', 'ShardingMatmul', 'ComputationCommunicationOverlap', 'KernelLaunch',
     'CublasBenchmark', 'CudnnBenchmark', 'GemmFlopsBenchmark', 'CudaGemmFlopsBenchmark', 'MemBwBenchmark',
     'CudaMemBwBenchmark', 'DiskBenchmark', 'IBLoopbackBenchmark', 'CudaNcclBwBenchmark', 'RocmMemBwBenchmark',
-    'RocmGemmFlopsBenchmark', 'GpuSmCopyBwBenchmark'
+    'RocmGemmFlopsBenchmark', 'GpuSmCopyBwBenchmark', 'IBTrafficBenchmark'
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/__init__.py:28]: IBBenchmark directly.

Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/__init__.py]: updated

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            name (str): benchmark name.
+            parameters (str): benchmark parameters.
+        """
+        super().__init__(name, parameters)
+
+        self._bin_name = 'ib_mpi'
+        self.__support_ib_commands = [
+            'ib_write_bw', 'ib_read_bw', 'ib_send_bw', 'ib_write_lat', 'ib_read_lat', 'ib_send_lat'
+        ]
+        self.__modes = ['one-to-one', 'one-to-many', 'many-to-one']
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py]: updated

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        self.__config_path = 'config.txt'
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py]: is current workdir ok? or /tmp?

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+            help='The path of config file on the target machine',
+        )
+        self._parser.add_argument(
+            '--bidirectional', action='store_true', default=False, help='Measure bidirectional bandwidth.'
+        )
+        self._parser.add_argument(
+            '--use_cuda', type=int, default=None, required=False, help='Test Use GPUDirect with the gpu index.'
+        )
+        self._parser.add_argument(
+            '--use_rocm', type=int, default=None, required=False, help='Test Use GPUDirect with the gpu index.'
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py]: updated

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            else:
+                ib_command_prefix = '{command} {command_params}'.format(
+                    command=ib_command, command_params=command_params
+                )
+                if 'bw' in ib_command and self._args.bidirectional:
+                    ib_command_prefix += ' -b'
+
+                command = os.path.join(self._args.bin_dir, self._bin_name)
+                command += ' --cmd_prefix ' + '\"' + ib_command_prefix + '\"'
+                command += ' --input_config ' + self.__config_path
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:250]: yes, I forgot that, added

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+            self._result.set_return_code(ReturnCode.MPI_INIT_FAILURE)
+            logger.error('No MPI environment - benchmark: {}.'.format(self._name))
+            return False
+        # Generate and check config
+        try:
+            if self._args.config is None:
+                self.gen_traffic_pattern(node_num, self._args.pattern, self.__config_path)
+            else:
+                self.__config_path = self._args.config
+            with open(self.__config_path, 'r') as f:
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:185]: per process

File: tests/benchmarks/micro_benchmarks/test_ib_traffic_performance.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+        # Test without ib devices
+        # Check registry.
+        benchmark_name = 'ib-traffic'
+        (benchmark_class,
+         predefine_params) = BenchmarkRegistry._BenchmarkRegistry__select_benchmark(benchmark_name, Platform.CPU)
+        assert (benchmark_class)
+
+        # Check preprocess
+        # Negative case
```
[the remark this answers is not in the document]
Comment (yukirora) [reply tests/benchmarks/micro_benchmarks/test_ib_traffic_performance.py]: updated

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+            '--ib_index',
+            type=int,
+            default=0,
+            required=False,
+            help='The index of ib device.',
+        )
+        self._parser.add_argument(
+            '--iters',
+            type=int,
+            default=20000,
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py]: updated to default as 5000

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        )
+        self._parser.add_argument(
+            '--commands',
+            type=str,
+            nargs='+',
+            default=['ib_write_bw'],
+            help='The ib command used to run, e.g., {}.'.format(' '.join(self.__support_ib_commands)),
+        )
+        self._parser.add_argument(
+            '--gid_index',
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:67]: some machines need to specify gid for ib test
https://github.com/linux-rdma/perftest#:~:text=default%3A%201000)%0A%20%20-x%2C-,--gid-index,-%3D%3Cindex%3E%09%09Test%20uses

File: superbench/benchmarks/return_code.py [hunk 10]
```diff
@@ -29,6 +29,7 @@ class ReturnCode(Enum):
     MICROBENCHMARK_RESULT_PARSING_FAILURE = 33
     MICROBENCHMARK_UNSUPPORTED_ARCHITECTURE = 34
     MICROBENCHMARK_DEVICE_GETTING_FAILURE = 35
+    MPI_INIT_FAILURE = 36
```
Comment (cp5555) [inline superbench/benchmarks/return_code.py:32]: Please follow the naming method as before. It's should be MICROBENCHMARK_MPI_INIT_FAILURE.

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+            config = self.__one_to_many(n)
+        elif mode == 'many-to-one':
+            config = self.__many_to_one(n)
+        elif mode == 'one-to-one':
+            config = self.__fully_one_to_one(n)
+        with open(config_file_path, 'w') as f:
+            for line in config:
+                f.write(line + '\n')
+
+    def _preprocess(self):    # noqa: C901
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:170]: Is it possible to optimize the code a little bit? There is too many # noqa: C901.

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+        for _ in range(count - 1):
+            pairs = [[robin[i], robin[count - i - 1]] for i in range(0, (count + 1) // 2)]
+            res.append(pairs)
+            robin = robin[1:]
+            robin = first + robin[1:] + robin[:1]
+        return res
+
+    def __fully_one_to_one(self, n):
+        config = []
+        if n % 2 == 1:
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:139]: The design is really strange. If failure happens, we will output a empty file, but we don't have any error message. People will think it's our code bug.

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+            return config
+        candidates = list(range(n))
+        res = self.__roundRobin(candidates)
+        for line in res:
+            row = []
+            for pair in line:
+                row.append('{},{}'.format(pair[0], pair[1]))
+            row = ';'.join(row)
+            config.append(row)
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:148]: I think we can append directly in the `__roundRobin()` and remove `__roundRobin()` function directly. What's the benefit for this design? Maybe I don't get the key idea on this.

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+            row = []
+            for client in range(n):
+                if server != client:
+                    pair = '{},{}'.format(server, client)
+                    row.append(pair)
+            row = ';'.join(row)
+            config.append(row)
+        return config
+
+    def __roundRobin(self, candidates):
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:125]: First, please write the comments for this algorithm.

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+        res = []
+        count = len(candidates)
+        robin = candidates
+        first = [candidates[0]]
+        for _ in range(count - 1):
+            pairs = [[robin[i], robin[count - i - 1]] for i in range(0, (count + 1) // 2)]
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:131]: Doesn't get it. `count %2 == 0`, why do we implement `(count + 1) // 2)`?

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py (hunk 14 above)
Comment (cp5555) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:125]: I think your implementation is circle method for Round-robin tournament (https://en.wikipedia.org/wiki/Round-robin_tournament)

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py (hunk 12 above)
Comment (cp5555) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:139]: Besides, circle method for Round-robin tournament can solve this issue by adding a fake IP. I don't know why we do in this way.

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+                    row.append(pair)
+            row = ';'.join(row)
+            config.append(row)
+        return config
+
+    def __roundRobin(self, candidates):
+        res = []
+        count = len(candidates)
+        robin = candidates
+        first = [candidates[0]]
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:129]: Please follow the algorithm definition. `first` is really weird.

**Review by cp5555 — APPROVED** [review]

File: superbench/benchmarks/micro_benchmarks/ib_validation_performance.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (abuccts) [reply superbench/benchmarks/micro_benchmarks/ib_validation_performance.py:67]: I think gid only works for RoCE, not InfiniBand

**Review by abuccts — APPROVED** [review]

**Pull request merged**